# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd
import numpy as np
import warnings
pd.set_option('display.max_columns', None) # display all columns
warnings.filterwarnings('ignore') # ignore warnings


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [2]:
# Your code here
austin=pd.read_csv('austin_weather.csv')
austin

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
...          ...        ...       ...       ...           ...          ...   
1314  2017-07-27        103        89        75            71           67   
1315  2017-07-28        105        91        76            71           64   
1316  2017-07-29        107        92        77            72           64   
1317  2017-07-30        106        93        79            70           68   
1318  2017-07-31         99        88        77            66           61   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

     SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                         29.86                     29.68   
1                         30.41                     30.13   
2                         30.56                     30.49   
3                         30.56                     30.45   
4                         30.41                     30.33   
...                         ...                       ...   
1314                      30.04                     29.97   
1315                      29.97                      29.9   
1316                      29.91                     29.86   
1317                      29.96                     29.91   
1318                      30.04                     29.97   

     SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                        29.59                  10                  7   
1                        29.87                  10                 10   
2                        30.41                  10                 10   
3                         30.3                  10                 10   
4                        30.27                  10                 10   
...                        ...                 ...                ...   
1314                     29.88                  10                 10   
1315                     29.81                  10                 10   
1316                     29.79                  10                 10   
1317                     29.87                  10                 10   
1318                     29.91                  10                 10   

     VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                     2          20          4          31   
1                     5          16          6          25   
2                    10           8          3          12   
3                     7          12          4          20   
4                     7          10          2          16   
...                 ...         ...        ...         ...   
1314                 10          12          5          21   
1315                

In [3]:
austin.columns = [column.lower() for column in austin.columns]

In [4]:
austin

date  temphighf  tempavgf  templowf dewpointhighf dewpointavgf  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
...          ...        ...       ...       ...           ...          ...   
1314  2017-07-27        103        89        75            71           67   
1315  2017-07-28        105        91        76            71           64   
1316  2017-07-29        107        92        77            72           64   
1317  2017-07-30        106        93        79            70           68   
1318  2017-07-31         99        88        77            66           61   

     dewpointlowf humidityhighpercent humidityavgpercent humiditylowpercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

     sealevelpressurehighinches sealevelpressureavginches  \
0                         29.86                     29.68   
1                         30.41                     30.13   
2                         30.56                     30.49   
3                         30.56                     30.45   
4                         30.41                     30.33   
...                         ...                       ...   
1314                      30.04                     29.97   
1315                      29.97                      29.9   
1316                      29.91                     29.86   
1317                      29.96                     29.91   
1318                      30.04                     29.97   

     sealevelpressurelowinches visibilityhighmiles visibilityavgmiles  \
0                        29.59                  10                  7   
1                        29.87                  10                 10   
2                        30.41                  10                 10   
3                         30.3                  10                 10   
4                        30.27                  10                 10   
...                        ...                 ...                ...   
1314                     29.88                  10                 10   
1315                     29.81                  10                 10   
1316                     29.79                  10                 10   
1317                     29.87                  10                 10   
1318                     29.91                  10                 10   

     visibilitylowmiles windhighmph windavgmph windgustmph  \
0                     2          20          4          31   
1                     5          16          6          25   
2                    10           8          3          12   
3                     7          12          4          20   
4                     7          10          2          16   
...                 ...         ...        ...         ...   
1314                 10          12          5          21   
1315                

In [5]:
austin.dewpointhighf.value_counts()

dewpointhighf
74    112
75     95
73     80
67     54
70     54
     ... 
22      1
80      1
26      1
25      1
13      1
Name: count, Length: 64, dtype: int64

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [6]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   date                        1319 non-null   object
 1   temphighf                   1319 non-null   int64 
 2   tempavgf                    1319 non-null   int64 
 3   templowf                    1319 non-null   int64 
 4   dewpointhighf               1319 non-null   object
 5   dewpointavgf                1319 non-null   object
 6   dewpointlowf                1319 non-null   object
 7   humidityhighpercent         1319 non-null   object
 8   humidityavgpercent          1319 non-null   object
 9   humiditylowpercent          1319 non-null   object
 10  sealevelpressurehighinches  1319 non-null   object
 11  sealevelpressureavginches   1319 non-null   object
 12  sealevelpressurelowinches   1319 non-null   object
 13  visibilityhighmiles         1319 non-null   obje

In [7]:
# Your code here
austin.describe().T

count       mean        std   min   25%   50%   75%    max
temphighf  1319.0  80.862775  14.766523  32.0  72.0  83.0  92.0  107.0
tempavgf   1319.0  70.642911  14.045904  29.0  62.0  73.0  83.0   93.0
templowf   1319.0  59.902957  14.190648  19.0  49.0  63.0  73.0   81.0

In [8]:
# Your code here
austin.head()

date  temphighf  tempavgf  templowf dewpointhighf dewpointavgf  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  dewpointlowf humidityhighpercent humidityavgpercent humiditylowpercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  sealevelpressurehighinches sealevelpressureavginches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  sealevelpressurelowinches visibilityhighmiles visibilityavgmiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  visibilitylowmiles windhighmph windavgmph windgustmph  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  precipitationsuminches               events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [9]:
# Your code here
austin.dewpointhighf.unique()

array(['67', '43', '31', '36', '44', '39', '41', '49', '51', '37', '55',
       '54', '15', '29', '60', '64', '58', '28', '42', '22', '38', '45',
       '19', '59', '62', '40', '17', '53', '34', '48', '52', '61', '63',
       '33', '57', '65', '46', '32', '50', '27', '56', '47', '68', '69',
       '66', '70', '73', '71', '76', '72', '75', '74', '-', '77', '23',
       '35', '30', '79', '78', '80', '26', '18', '25', '13'], dtype=object)

### Your observation here:
the data has objcet dtype because of the presence of non-numeric values (special characters)within the data.


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [10]:
df=austin.copy()

In [11]:
wrong_type_columns = ['dewpointhighf', 'dewpointavgf', 'dewpointlowf', 'humidityhighpercent', 
                      'humidityavgpercent', 'humiditylowpercent', 'sealevelpressurehighinches', 
                      'sealevelpressureavginches' ,'sealevelpressurelowinches', 'visibilityhighmiles',
                      'visibilityavgmiles', 'visibilitylowmiles', 'windhighmph', 'windavgmph', 
                      'windgustmph', 'precipitationsuminches']

In [12]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [13]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   date                        1319 non-null   object 
 1   temphighf                   1319 non-null   int64  
 2   tempavgf                    1319 non-null   int64  
 3   templowf                    1319 non-null   int64  
 4   dewpointhighf               1312 non-null   float64
 5   dewpointavgf                1312 non-null   float64
 6   dewpointlowf                1312 non-null   float64
 7   humidityhighpercent         1317 non-null   float64
 8   humidityavgpercent          1317 non-null   float64
 9   humiditylowpercent          1317 non-null   float64
 10  sealevelpressurehighinches  1316 non-null   float64
 11  sealevelpressureavginches   1316 non-null   float64
 12  sealevelpressurelowinches   1316 non-null   float64
 13  visibilityhighmiles         1307 

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [14]:
# Your code here
missing_values =austin[austin.isnull().any(axis=1)]
missing_values

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
4     2013-12-25         58        50        41           44.0          40.0   
6     2013-12-27         60        53        45           41.0          39.0   
7     2013-12-28         62        51        40           43.0          39.0   
42    2014-02-01         76        66        55           62.0          59.0   
51    2014-02-10         60        48        35           49.0          36.0   
...          ...        ...       ...       ...            ...           ...   
1269  2017-06-12         94        85        75           72.0          68.0   
1288  2017-07-01         98        89        79           76.0          72.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1308  2017-07-21        104        91        77           74.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      humiditylowpercent  sealevelpressurehighinches  \
4                   56.0                       30.41   
6                   47.0                       30.46   
7                   36.0                       30.33   
42                  60.0                       29.91   
51                  66.0                       30.23   
...                  ...                         ...   
1269                36.0                       29.96   
1288                38.0                       30.13   
1304                38.0                       29.94   
1308                27.0                       30.01   
1313                25.0                       30.13   

      sealevelpressureavginches  sealevelpressurelowinches  \
4                         30.33                      30.27   
6                         30.39                      30.34   
7                         30.17                      30.04   
42                        29.81                      29.75   
51                        30.15                      30.02   
...                         ...                        ...   
1269                      29.89                      29.81   
1288                      30.02                      29.92   
1304                      29.90                      29.84   
1308                      29.95                      29.86   
1313                      30.03                      29.93   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
4                    10.0                10.0                 7.0   
6                    10.0                 9.0                 7.0   
7                    10.0                10.0                 7.0   
42                   10.0                10.0                 9.0   
51                   10.0                 8.0                 4.0   
...                   ...                 ...                 ...   
1269                 10.0                10.0                 2.0   
1288                 10.0                 9.0                 6.0   
1304                 10.0                10.0                10.0   
1308                 10.0                10.0                10.0   
1313                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph  pre

In [15]:
#Extra code : 
missing_values.isnull().sum()


date                            0
temphighf                       0
tempavgf                        0
templowf                        0
dewpointhighf                   7
dewpointavgf                    7
dewpointlowf                    7
humidityhighpercent             2
humidityavgpercent              2
humiditylowpercent              2
sealevelpressurehighinches      3
sealevelpressureavginches       3
sealevelpressurelowinches       3
visibilityhighmiles            12
visibilityavgmiles             12
visibilitylowmiles             12
windhighmph                     2
windavgmph                      2
windgustmph                     4
precipitationsuminches        124
events                          0
dtype: int64

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

In [16]:
missing_values_per_row = austin.isnull().sum(axis=1)

threshold_ratio = 0.5 
most_missing_rows = austin[missing_values_per_row / austin.shape[1] > threshold_ratio]

austin_cleaned = austin.drop(index=most_missing_rows.index)

austin_cleaned = austin_cleaned.interpolate(method='linear')
austin_cleaned


date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph  pre

#### Next, count the number of rows of `austin` and `missing_values`.

In [17]:
austin_rows = austin.shape[0]

# Count the number of rows in the 'missing_values' DataFrame
missing_values_rows = missing_values.shape[0]
missing_values_rows

136

In [18]:
missing_values.info()

<class 'pandas.core.frame.DataFrame'>
Index: 136 entries, 4 to 1313
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   date                        136 non-null    object 
 1   temphighf                   136 non-null    int64  
 2   tempavgf                    136 non-null    int64  
 3   templowf                    136 non-null    int64  
 4   dewpointhighf               129 non-null    float64
 5   dewpointavgf                129 non-null    float64
 6   dewpointlowf                129 non-null    float64
 7   humidityhighpercent         134 non-null    float64
 8   humidityavgpercent          134 non-null    float64
 9   humiditylowpercent          134 non-null    float64
 10  sealevelpressurehighinches  133 non-null    float64
 11  sealevelpressureavginches   133 non-null    float64
 12  sealevelpressurelowinches   133 non-null    float64
 13  visibilityhighmiles         124 non-nul

#### Calculate the ratio of missing rows to total rows

In [19]:
missing_rows_ratio = missing_values_rows / austin_rows
missing_rows_ratio

0.10310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [20]:
# Your code here
missing_rows_per_column = austin.isna().sum()
missing_rows_per_column

date                            0
temphighf                       0
tempavgf                        0
templowf                        0
dewpointhighf                   7
dewpointavgf                    7
dewpointlowf                    7
humidityhighpercent             2
humidityavgpercent              2
humiditylowpercent              2
sealevelpressurehighinches      3
sealevelpressureavginches       3
sealevelpressurelowinches       3
visibilityhighmiles            12
visibilityavgmiles             12
visibilitylowmiles             12
windhighmph                     2
windavgmph                      2
windgustmph                     4
precipitationsuminches        124
events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [21]:
# Your code here
missing_values_precipitation = missing_rows_per_column['precipitationsuminches']

ratio_missing_precipitation = missing_values_precipitation / austin.shape[0]
ratio_missing_precipitation

0.09401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [22]:
# Your code here 
austin.drop(columns='precipitationsuminches', inplace=True)

# Print `austin` to confirm the column is indeed removed

austin

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph     

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [23]:
# Your code here
austin_fixed = austin.interpolate(method='linear', inplace=False)
austin_fixed 

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph     

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [24]:
# Your code here
austin.isnull().sum()


date                           0
temphighf                      0
tempavgf                       0
templowf                       0
dewpointhighf                  7
dewpointavgf                   7
dewpointlowf                   7
humidityhighpercent            2
humidityavgpercent             2
humiditylowpercent             2
sealevelpressurehighinches     3
sealevelpressureavginches      3
sealevelpressurelowinches      3
visibilityhighmiles           12
visibilityavgmiles            12
visibilitylowmiles            12
windhighmph                    2
windavgmph                     2
windgustmph                    4
events                         0
dtype: int64

In [25]:
missing_fixed = austin_fixed.isnull().sum()
missing_fixed

date                          0
temphighf                     0
tempavgf                      0
templowf                      0
dewpointhighf                 0
dewpointavgf                  0
dewpointlowf                  0
humidityhighpercent           0
humidityavgpercent            0
humiditylowpercent            0
sealevelpressurehighinches    0
sealevelpressureavginches     0
sealevelpressurelowinches     0
visibilityhighmiles           0
visibilityavgmiles            0
visibilitylowmiles            0
windhighmph                   0
windavgmph                    0
windgustmph                   0
events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [26]:
# Your code here:
events_counts = austin_fixed['events'].value_counts()
events_counts

events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [27]:
# Your answer:
max_events_per_day = austin_fixed['events'].str.split(',').apply(lambda x: len(x)).max()
max_events_per_day

3

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [28]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph     

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [29]:
# Your code here
for event in event_list:
    austin_fixed[event] = austin_fixed['events'].str.contains(event).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [30]:
# Your code here
austin_fixed

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph     

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [31]:
# Your code here
austin_fixed.drop(columns='events', inplace=True)

In [33]:
austin_fixed

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph  Sno

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [34]:
# Your code here
from datetime import datetime

austin_fixed['date'] = austin_fixed['date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d').toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [35]:
austin_fixed

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1314  736537        103        89        75           71.0          67.0   
1315  736538        105        91        76           71.0          64.0   
1316  736539        107        92        77           72.0          64.0   
1317  736540        106        93        79           70.0          68.0   
1318  736541         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph  Snow  Fog  Rain  Thunderstorm  
0            20

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [39]:
# Your code here:
X = austin_fixed[['date',  'temphighf',  'templowf',  'dewpointhighf',  'dewpointavgf',  
                  'dewpointlowf',  'humidityhighpercent',  'humidityavgpercent',  'humiditylowpercent',  
                  'sealevelpressurehighinches',  'sealevelpressureavginches',  'sealevelpressurelowinches',  
                  'visibilityhighmiles',  'visibilityavgmiles',  'visibilitylowmiles',  'windhighmph',  
                  'windavgmph',  'windgustmph',  'Snow',  'Fog',  'Rain',  'Thunderstorm']]

y = austin_fixed[['tempavgf']]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [40]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [41]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [42]:
X_train

date  temphighf  templowf  dewpointhighf  dewpointavgf  dewpointlowf  \
598   735821        105        75           70.0         69.25          67.5   
1213  736436         80        63           67.0         65.00          57.0   
209   735432         78        72           74.0         71.00          68.0   
538   735761         93        75           75.0         73.00          71.0   
140   735363         91        60           66.0         62.00          58.0   
...      ...        ...       ...            ...           ...           ...   
1095  736318         59        35           25.0         20.00          15.0   
1130  736353         84        50           64.0         51.00          36.0   
1294  736517         99        75           74.0         71.00          65.0   
860   736083         84        75           74.0         73.00          72.0   
1126  736349         76        47           55.0         50.00          46.0   

      humidityhighpercent  humidityavgpercent  humiditylowpercent  \
598                  77.5                52.5                27.5   
1213                 90.0                80.0                69.0   
209                 100.0                87.0                74.0   
538                  88.0                70.0                52.0   
140                  97.0                71.0                44.0   
...                   ...                 ...                 ...   
1095                 59.0                40.0                20.0   
1130                 87.0                62.0                37.0   
1294                 88.0                61.0                33.0   
860                  94.0                82.0                70.0   
1126                 93.0                66.0                38.0   

      sealevelpressurehighinches  sealevelpressureavginches  \
598                      29.9725                    29.9275   
1213                     30.0500                    30.0000   
209                      30.0100                    29.9300   
538                      29.8700                    29.8200   
140                      29.9200                    29.8400   
...                          ...                        ...   
1095                     30.5900                    30.4600   
1130                     29.8000                    29.7100   
1294                     30.1300                    30.0500   
860                      29.7800                    29.6500   
1126                     29.8400                    29.7400   

      sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
598                      29.895                 10.0                10.0   
1213                     29.930                 10.0                10.0   
209                      29.870                 10.0                 7.0   
538                      29.760                 10.0                10.0   
140                      29.750                 10.0                10.0   
...                         ...                  ...                 ...   
1095                     30.300                 10.0                10.0   
1130                     29.610                 10.0                10.0   
1294                     29.980                 10.0                10.0   
860                      29.540                 10.0                 7.0   
1126                     29.650                 10.0                 9.0   

      visibilitylowmiles  windhighmph  windavgmph  windgustmph  Snow  Fog  \
598                 10.0         13.5         4.5         20.0     0    0   
1213                 6.0         16.0         5.0         26.0     0    0   
209                  1.0         14.0         3.0         27.0     0    0   
538                 10.0         13.0         6.0         20.0     0    0   
140                  5.0         16.0         6.0         23.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1095     

In [43]:
X_test

date  temphighf  templowf  dewpointhighf  dewpointavgf  dewpointlowf  \
677   735900         81        51           64.0          54.0          49.0   
1046  736269         91        71           73.0          71.0          64.0   
610   735833        101        76           76.0          72.0          65.0   
49    735272         65        37           42.0          36.0          29.0   
1284  736507         91        71           74.0          72.0          67.0   
...      ...        ...       ...            ...           ...           ...   
1176  736399         66        61           64.0          62.0          59.0   
1002  736225         98        75           75.0          73.0          71.0   
1159  736382         87        51           49.0          45.0          40.0   
542   735765         83        73           75.0          73.0          70.0   
170   735393         86        68           75.0          71.0          66.0   

      humidityhighpercent  humidityavgpercent  humiditylowpercent  \
677                  96.0                66.0                35.0   
1046                100.0                72.0                44.0   
610                  94.0                64.0                33.0   
49                   85.0                63.0                40.0   
1284                100.0                75.0                50.0   
...                   ...                 ...                 ...   
1176                100.0                94.0                87.0   
1002                 94.0                70.0                45.0   
1159                 71.0                48.0                24.0   
542                  94.0                84.0                74.0   
170                  91.0                78.0                65.0   

      sealevelpressurehighinches  sealevelpressureavginches  \
677                        29.97                      29.89   
1046                       30.04                      29.98   
610                        30.05                      29.97   
49                         30.26                      30.17   
1284                       30.05                      30.00   
...                          ...                        ...   
1176                       30.14                      30.09   
1002                       29.99                      29.93   
1159                       29.96                      29.86   
542                        29.96                      29.90   
170                        29.92                      29.81   

      sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
677                       29.83                 10.0                10.0   
1046                      29.90                 10.0                 9.0   
610                       29.91                 10.0                10.0   
49                        30.03                 10.0                 7.0   
1284                      29.90                 10.0                 7.0   
...                         ...                  ...                 ...   
1176                      30.01                 10.0                 6.0   
1002                      29.89                 10.0                10.0   
1159                      29.73                 10.0                10.0   
542                       29.78                 10.0                 8.0   
170                       29.75                 10.0                 8.0   

      visibilitylowmiles  windhighmph  windavgmph  windgustmph  Snow  Fog  \
677                 10.0         12.0         4.0         18.0     0    0   
1046                 1.0         15.0         5.0         22.0     0    0   
610                 10.0         12.0         5.0         18.0     0    0   
49                   3.0         14.0         5.0         21.0     0    0   
1284                 2.0         20.0         3.0         30.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1176     

In [44]:
y_train

tempavgf
598         90
1213        72
209         75
538         84
140         76
...        ...
1095        47
1130        67
1294        87
860         80
1126        62

[1055 rows x 1 columns]

In [45]:
y_test

tempavgf
677         66
1046        81
610         89
49          51
1284        81
...        ...
1176        64
1002        87
1159        69
542         78
170         77

[264 rows x 1 columns]

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
